instaloader/instaloader.py

```diff
@@ -312,15 +312,23 @@ def download_pic(self, filename: str, url: str, mtime: datetime,
                      filename_suffix: Optional[str] = None, _attempt: int = 1) -> bool:
         """Downloads and saves picture with given url under given directory with given timestamp.
         Returns true, if file was actually downloaded, i.e. updated."""
-        urlmatch = re.search('\\.[a-z0-9]*\\?', url)
-        file_extension = url[-3:] if urlmatch is None else urlmatch.group(0)[1:-1]
         if filename_suffix is not None:
             filename += '_' + filename_suffix
-        filename += '.' + file_extension
+        urlmatch = re.search('\\.[a-z0-9]*\\?', url)
+        file_extension = url[-3:] if urlmatch is None else urlmatch.group(0)[1:-1]
+        nominal_filename = filename + '.' + file_extension
+        if os.path.isfile(nominal_filename):
+            self.context.log(nominal_filename + ' exists', end=' ', flush=True)
+            return False
+        resp = self.context.get_raw(url)
+        if 'Content-Type' in resp.headers and resp.headers['Content-Type']:
+            header_extension = '.' + resp.headers['Content-Type'].split(';')[0].split('/')[-1]
+            header_extension = header_extension.lower().replace('jpeg', 'jpg')
+            filename += header_extension
         if os.path.isfile(filename):
             self.context.log(filename + ' exists', end=' ', flush=True)
             return False
```

Comment:
This is actually only needed if a different file extension was determined using the Content-Type header.

Reply:
I thought about it; but if the file with the same extension already existed, we have already returned before. 

So while is kinda redundant, it would help in some rare cases the file became existing during the first check and after the request is made. 

What do you think?

Reply:
Anyway, I modified it so it will only check if the ext is different.